pkg/operator/agent/agent.go

```diff
@@ -217,40 +221,68 @@ func (a *Agent) discoverFlexvolumeDir() string {
 		return getDefaultFlexvolumeDir()
 	}
 
-	// unmarshall to a NodeConfigKubelet
+	// unmarshal to a NodeConfigKubelet
 	configKubelet := NodeConfigKubelet{}
 	if err := json.Unmarshal(nodeConfig, &configKubelet); err != nil {
 		logger.Warningf("unable to parse node config from Kubelet: %+v", err)
-		return getDefaultFlexvolumeDir()
+	} else {
+		flexvolumeDirPath = configKubelet.ComponentConfig.VolumePluginDir
 	}
-	flexvolumeDirPath = configKubelet.ComponentConfig.VolumePluginDir
-	if flexvolumeDirPath == "" {
-		// unmarshall to a NodeConfigControllerManager
-		configControllerManager := NodeConfigControllerManager{}
-		if err := json.Unmarshal(nodeConfig, &configControllerManager); err != nil {
-			logger.Warningf("unable to parse node config from controller manager: %+v", err)
-			return getDefaultFlexvolumeDir()
-		}
+
+	if flexvolumeDirPath != "" {
```

Comment:
why check for a value here? Why not simply log and return in the previous code block on line 229? Seems like the checks for empty string could be removed in that case. If the unmarshal fails, have you seen that it can actually successfully assign a value to the path?

Reply:
there are 3 types we know of that can be returned from the node config API.  if the JSON response is of a different type than we attempt to unmarshal it as, the unmarshal operation succeeds but we get the zero values for the fields in the unmarshalled struct.  JSON decoding in golang is very permissive in that way.  See this example: https://play.golang.org/p/VbhEoylRED

Just because the unmarshal did not return an error, that does not in any way guarantee that we will have a non empty flexvolume directory path.  That's why i'm checking after each unmarshal to see if we actually have an answer, because we often don't, and we then need to move onto the next try.

i have not seen the unmarshal straight up fail before, that may only happen if the JSON is malformed, but i'm not certain.  I was trying to be as resilient as possible such that if unmarshalling returns an error we will try with the next type.

Reply:
ok, i see that it might not even fail, but it could still be empty